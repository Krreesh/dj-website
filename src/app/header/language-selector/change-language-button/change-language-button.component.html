<button class="dropdown-item changeLanguage"
        data-change-language-button
        (click)="useLanguage(language)">{{language | uppercase}}</button>
