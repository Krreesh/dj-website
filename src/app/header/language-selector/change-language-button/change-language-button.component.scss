@import "../../../../colors";

.dropdown-item {
  padding: .25rem 1.25rem;
}
