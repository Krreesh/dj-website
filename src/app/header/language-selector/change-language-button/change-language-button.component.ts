import {Component, Input} from '@angular/core';
import {LanguageService} from '../../../translation/language.service';

@Component({
  selector: 'app-change-language-button',
  styleUrls: ['./change-language-button.component.scss'],
  templateUrl: './change-language-button.component.html'
})
export class ChangeLanguageButtonComponent {

  @Input()
  language: string;

  constructor(private languageService: LanguageService) {
  }

  useLanguage(language: string): void {
    this.languageService.updateCurrentLanguage(language);
  }
}
