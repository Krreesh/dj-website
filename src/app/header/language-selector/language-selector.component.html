<ng-container *ngIf="currentAndOtherLanguage$ | async as languages">
  <div ngbDropdown class="d-inline-block dropdown">
    <button class="btn btn-outline-dark dropdown-toggle changeLanguageButton" id="languageSelectorDropdown"
            ngbDropdownToggle>{{languages.current | uppercase}}</button>
    <div ngbDropdownMenu aria-labelledby="languageSelectorDropdown" class="dropdown-menu">
      <app-change-language-button *ngFor="let otherLanguage of languages.other" [language]="otherLanguage"></app-change-language-button>
    </div>
  </div>
</ng-container>
