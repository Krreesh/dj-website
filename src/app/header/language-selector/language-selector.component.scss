@import "../../../styles";

.dropdown-menu {
  min-width: 0;
}

#languageSelectorDropdown {
  color: $black-color;
}

.changeLanguageButton{
  font-size: 1.2rem;
}
