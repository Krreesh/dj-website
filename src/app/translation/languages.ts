export class Languages {
  readonly current: string;
  readonly other: string[];

  private constructor(currentLanguage: string, supportedLanguages: string[]) {
    this.current = currentLanguage;
    this.other = supportedLanguages.filter(
      (supportedLanguage) => supportedLanguage !== currentLanguage
    );
  }

  static from(
    currentLanguage: string,
    supportedLanguages: string[]
  ): Languages {
    Languages.assertNotEmpty(supportedLanguages);
    Languages.assertIncludes(supportedLanguages, currentLanguage);
    return new Languages(currentLanguage, supportedLanguages);
  }

  private static assertNotEmpty(supportedLanguages: string[]) {
    if (supportedLanguages.length === 0) {
      throw new Error(
        'Invalid argument exception: Supported languages cannot be empty.'
      );
    }
  }

  private static assertIncludes(
    supportedLanguages: string[],
    currentLanguage: string
  ) {
    if (!supportedLanguages.includes(currentLanguage)) {
      throw new Error(
        'Invalid argument exception: Supported languages should include the current language.'
      );
    }
  }
}
