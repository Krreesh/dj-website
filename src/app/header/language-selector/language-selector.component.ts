import {Component} from '@angular/core';
import {LanguageService} from '../../translation/language.service';
import {Observable} from 'rxjs';
import {Languages} from '../../translation/languages';

@Component({
  selector: 'app-language-selector',
  templateUrl: './language-selector.component.html',
  styleUrls: ['./language-selector.component.scss']
})
export class LanguageSelectorComponent {
  currentAndOtherLanguage$: Observable<Languages>;

  constructor(private languageService: LanguageService) {
    this.currentAndOtherLanguage$ = this.languageService.getCurrentAndOtherLanguages$();
  }
}
