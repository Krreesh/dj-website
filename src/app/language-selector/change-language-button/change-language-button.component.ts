import {Component, Input} from '@angular/core';
import {LanguageService} from '../../translation/language-service';

@Component({
  selector: 'app-change-language-button',
  template: `
    <button class="dropdown-item" ngbDropdownItem (click)="useLanguage(language)">
      {{language}}
    </button>
  `,
  styles: []
})
export class ChangeLanguageButtonComponent {

  @Input()
  language: string;

  constructor(private languageService: LanguageService) {
  }

  useLanguage(language: string): void {
    this.languageService.setCurrentLanguage(language);
  }
}
