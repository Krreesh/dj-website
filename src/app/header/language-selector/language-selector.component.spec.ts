import {ComponentFixture, TestBed} from '@angular/core/testing';
import {LanguageSelectorComponent} from './language-selector.component';
import {LanguageService} from '../../translation/language.service';
import {ChangeLanguageButtonComponent} from './change-language-button/change-language-button.component';
import {BehaviorSubject} from 'rxjs';
import {Languages} from '../../translation/languages';
import {Builder} from 'builder-pattern';

describe('LanguageSelectorComponent', () => {
  let component: LanguageSelectorComponent;
  let fixture: ComponentFixture<LanguageSelectorComponent>;
  let currentAndOtherLanguagesSubject = new BehaviorSubject<Languages>(null);

  beforeEach(() => {
    const currentAndOtherLanguages$ = jest.fn().mockReturnValue(currentAndOtherLanguagesSubject.asObservable());

    TestBed.configureTestingModule({
      declarations: [
        LanguageSelectorComponent,
        ChangeLanguageButtonComponent
      ],
      providers: [{
        provide: LanguageService,
        useValue: {
          getCurrentAndOtherLanguages$: currentAndOtherLanguages$
        }
      }]
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(LanguageSelectorComponent);
    component = fixture.debugElement.componentInstance;
  });

  it('should render the current language and the other languages to choose from', () => {
    const languages = Builder<Languages>()
      .current('en')
      .other(['nl', 'fr'])
      .build();
    currentAndOtherLanguagesSubject.next(languages);

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
