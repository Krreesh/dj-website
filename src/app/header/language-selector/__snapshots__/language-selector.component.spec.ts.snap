// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelectorComponent should render the current language and the other languages to choose from 1`] = `
<app-language-selector
  currentAndOtherLanguage$={[Function Observable]}
  languageService={[Function Object]}
>
  <div
    class="d-inline-block dropdown"
    ngbdropdown=""
  >
    <button
      class="btn btn-outline-dark dropdown-toggle changeLanguageButton"
      id="languageSelectorDropdown"
      ngbdropdowntoggle=""
    >
      EN
    </button>
    <div
      aria-labelledby="languageSelectorDropdown"
      class="dropdown-menu"
      ngbdropdownmenu=""
    >
      
      <app-change-language-button>
        <button
          class="dropdown-item changeLanguage"
          data-change-language-button=""
        >
          NL
        </button>
      </app-change-language-button>
      <app-change-language-button>
        <button
          class="dropdown-item changeLanguage"
          data-change-language-button=""
        >
          FR
        </button>
      </app-change-language-button>
    </div>
  </div>
</app-language-selector>
`;
